File: elements-of-programming-interviews/cpp/sunset_view.cc
#include <iterator>
#include <vector>

#include "test_framework/generic_test.h"
using std::pair;
using std::vector;

/*
Cases:
  - Every building equal/ gets taller than previous; only last can view
  - Every building gets shorter; all can see it
  - Any time a taller building comes in, any previous shorter one can no longer
see and nothing changes this
  - If  building is shorter than top, push it.
  - If a building is equal or taller than top, pop until empty or we reach one
that's taller, then push.
O(n) space, O(n) time (evaluate, push, and pop almost every element)
*/

vector<int> ExamineBuildingsWithSunset(
    vector<int>::const_iterator sequence_begin,
    const vector<int>::const_iterator& sequence_end) {
  vector<pair<int, int>> s;
  int i = 0;
  for (auto it = sequence_begin; it != sequence_end; i++, it++) {
    if (s.empty() || s.back().first > *it) {
      s.push_back({*it, i});
    } else {
      while (!s.empty() && s.back().first <= *it) {
        s.pop_back();
      }
      s.push_back({*it, i});
    }
  }
  vector<int> final;
  for (auto it = s.rbegin(); it != s.rend(); it++) {
    final.push_back(it->second);
  }
  return final;
}
vector<int> ExamineBuildingsWithSunsetWrapper(const vector<int>& sequence) {
  return ExamineBuildingsWithSunset(cbegin(sequence), cend(sequence));
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"sequence"};
  return GenericTestMain(args, "sunset_view.cc", "sunset_view.tsv",
                         &ExamineBuildingsWithSunsetWrapper,
                         DefaultComparator{}, param_names);
}

File: elements-of-programming-interviews/python/sunset_view.py
from typing import Iterator, List

from test_framework import generic_test

"""
- Input is a list of height
- Output is list of indices

[6,5,4,3,2,1]
[1,2,3,4]
"""


def examine_buildings_with_sunset(sequence: Iterator[int]) -> List[int]:
    stack = []
    for i, height in enumerate(sequence):
        while stack and stack[-1][1] <= height:
            stack.pop()
        stack.append((i, height))
    return [item[0] for item in reversed(stack)]


def examine_buildings_with_sunset_wrapper(sequence):
    return examine_buildings_with_sunset(iter(sequence))


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('sunset_view.py', 'sunset_view.tsv',
                                       examine_buildings_with_sunset))
